### PR TITLE
Add RATS dataset for SV task

### PR DESCRIPTION
**Pull request by shimhz**
## What?

Add RATS dataset for SV task 

## Why?

The RATS dataset includes detailed channel information so that it could be useful for channel/codec analysis in SV task.

## See also

[(https://catalog.ldc.upenn.edu/LDC2021S08)](url)

### CODE REVIEW

**Comment by sw005320** [comment]
@Jungjee, can you review this PR?

**Comment by Jungjee** [comment]
> @Jungjee, can you review this PR?

sure!

File: egs2/rats/spk1/conf/train_RawNet3.yaml [hunk 1]
```diff
@@ -0,0 +1,91 @@
+# RawNet3 reproduce recipe configuration.
+# Requires approx. 40GRAM per GPU when run with 2GPUs
+# or 20GRAM when run with 4GPUs.
```
Comment (Jungjee) [inline egs2/rats/spk1/conf/train_RawNet3.yaml:3]: Dos this align with your actual training config? 
If not, can you reflect the setting you ran?

File: egs2/rats/spk1/local/data.sh [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    python local/data_prep.py --src "${RATS}data/dev/" --dst "${trg_dir}/rats_test/"
+    python local/data_prep.py --src "${RATS}data/train/" --dst "${trg_dir}/rats_train/"
+
+    for f in wav.scp utt2spk spk2utt; do
+        sort ${trg_dir}/rats_test/${f} -o ${trg_dir}/rats_test/${f}
+        sort ${trg_dir}/rats_train/${f} -o ${trg_dir}/rats_train/${f}
+    done
+
+    # make test trial compatible with ESPnet.
+    python local/convert_trial.py --trial rats_eval_trial.txt --scp ${trg_dir}/rats_test/wav.scp --src2spk ${trg_dir}/rats_test/src2spk --out ${trg_dir}/rats_test
```
Comment (Jungjee) [inline egs2/rats/spk1/local/data.sh:61]: can you add a comment on what `src2spk` does here?

File: egs2/rats/spk1/local/data_prep.py [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    for line in open(f_info, "r").readlines()[1:]:
+        src_id, lng, spk, t, part = line.strip().split()
+        src2spk[src_id] = spk
+    pk.dump(src2spk, open(dst + "src2spk", "wb"))
+
+    spk2utt = {}
+    utt2spk = []
+    wav_list = []
+
+    fs = glob.glob(os.path.join(src, "audio/*/*/*.flac"))
```
Comment (Jungjee) [inline egs2/rats/spk1/local/data_prep.py:24]: I think you're assuming that the RATs data is already downloaded as exists in a specific directory. 

I believe there are cases (especially when the data needs to be purchased) where we don't support the download in `data.sh`. 
However, in this case, please add a few lines of comments on how you need the dataset or filetree to be prepared).

File: egs2/rats/spk1/run.sh [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+set -e
+set -u
+set -o pipefail
+
+
+spk_config=conf/train_RawNet3.yaml
+
+train_set="rats_train"
+valid_set="rats_test"
+cohort_set="rats_test"
```
Comment (Jungjee) [inline egs2/rats/spk1/run.sh:11]: can you remove this line? I believe your experiment didn't use score normalization.
It is not ideal to note the test set as cohort set as you'd know.

File: egs2/rats/spk1/run.sh [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+test_sets="rats_test"
+feats_type="raw"
+
+./spk.sh \
+    --feats_type ${feats_type} \
+    --spk_config ${spk_config} \
+    --train_set ${train_set} \
+    --valid_set ${valid_set} \
+    --cohort_set ${cohort_set} \
```
Comment (Jungjee) [inline egs2/rats/spk1/run.sh:20]: ditto, recommend removing this line as well

**Review by Jungjee — COMMENTED** [review]
Hi @shimhz, thanks for adding a new recipe using ESPnet-SPK!

I left a few comments. One more thing, can you add a readme.md file under `egs/rats/spk1`? 
You can refer to what is in `egs/voxceleb/spk1/readme.md` to check which information is required.

File: egs2/rats/spk1/run.sh (hunk 4 above)
Comment (Jungjee) [reply egs2/rats/spk1/run.sh:11]: sorry, only L11

File: egs2/rats/spk1/run.sh (hunk 5 above)
Comment (Jungjee) [reply egs2/rats/spk1/run.sh:20]: sorry, only L20

**Comment by shimhz** [comment]
I reflected the comments except for README.file as I did not get the result yet. 
I'm going to include the result of RawNet3 in the README.file tomorrow.

**Comment by shimhz** [comment]
I've done the evaluation again, but it still showed over 50% of EER.

I've double-checked the data prep part but I couldn't find any problems from my side.
I think I need to touch the optimization part.

image


eer
loss

**Comment by sw005320** [comment]
Did you tune the learning rate or other optimization hyper-parameters?

**Comment by shimhz** [comment]
No, I used the same parameters as the Voxceleb config, but I should fine-tune it.

**Comment by shimhz** [comment]
I've updated the PR only to include source partition for the training and it showed 19.72% of EER.

File: egs2/rats/spk1/local/rats_trial_prep.py [hunk 6]
```diff
@@ -1,75 +1,104 @@
...
-            elif class_idx == 1:
-                while True:
-                    tmp_lines = list(set(l_dev) - set(d_spk2utt[spk]))
-                    utt1, utt2 = np.random.choice(tmp_lines, size=2, replace=False)
-                    if is_unique_trial(class_idx, utt1, utt2):
-                        break
-
-            f_out.write("{} {} {}\n".format(class_idx, utt1, utt2))
+        # Pre-generate the random speaker and class indices to reduce calls to random
+        n_trials = len(d_src2spk) * 20
```
Comment (Jungjee) [inline egs2/rats/spk1/local/rats_trial_prep.py:58]: How about moving this to the beginning part so that others could potentially be aware and maybe play with this variable?

File: egs2/rats/spk1/local/rats_trial_prep.py [hunk 7]
```diff
@@ -1,75 +1,104 @@
...
+        random_spks = np.random.choice(spk_keys, size=n_trials)
+        class_indices = np.random.randint(0, 2, size=n_trials)
+
+        batch_size = 100  # Process in batches
```
Comment (Jungjee) [inline egs2/rats/spk1/local/rats_trial_prep.py:62]: ditto, move along with n_trials?

File: egs2/rats/spk1/local/rats_trial_prep.py [hunk 8]
```diff
@@ -1,75 +1,104 @@
...
+            batch_trials = []
+
+            for spk, class_idx in zip(batch_spks, batch_classes):
+                if class_idx == 1:
+                    # Same-speaker pair
+                    utt_list = d_spk2utt[spk]
+                    if len(utt_list) >= 2:
+                        utt_pairs = np.random.choice(
+                            utt_list,
+                            size=(min(batch_size, len(utt_list) // 2), 2),
```
Comment (Jungjee) [inline egs2/rats/spk1/local/rats_trial_prep.py:76]: It's not wrong nor critical, but we could think about this part where it seems to use (almost) all utterances once for target pair.
Could it be better to use some utterances multiple times, paired with different utterances? (an open question, which I don't have answer to)

**Review by Jungjee — APPROVED** [review]
I have one minor suggestion and an open question, but otherwise, LGTM! Thanks, @shimhz

**Comment by Jungjee** [comment]
@shimhz, sorry, one more thing. Can you close and re-open this PR to activate the CIs?

File: egs2/rats/spk1/README.md [hunk 9]
```diff
@@ -0,0 +1,19 @@
+# RESULTS
```
Comment (sw005320) [inline egs2/rats/spk1/README.md:1]: Please upload the model on the HF hub and add the link here.

File: egs2/README.md
Comment (sw005320) [inline egs2/README.md:1]: I still see that coraal and ogi_kids_speech are deleted...
Please fix them.

Comment (shimhz) [reply egs2/README.md:1]: Oh sorry I just fixed them.

**Comment by sw005320** [comment]
Can you fix this error?
https://github.com/espnet/espnet/actions/runs/11801693175/job/32875629756?pr=5840#step:8:572

**Comment by sw005320** [comment]
Done!
Thanks!
